WordPress/gutenberg
# [Patterns] Separate sync status into a filter control

## What and Why?

Part of https://github.com/WordPress/gutenberg/issues/51948. Revamp the layout of the "My patterns" screen and refactor some code around it to better prepare for pagination in the future.

The goal is to limit the number of patterns rendered on screen so that users with many patterns (>100) won't get frozen or crash. This is a temporary solution until we decide which accessible pagination system to use.

## How?

Worth noting that even though we only render at most 100 patterns at a time, we're still fetching all the patterns from the API. It's needed so that we can do client-side filtering and searching. In the future, we might decide to switch to server-side filtering and searching (#52396) once the API supports them.

## Testing Instructions




1. Go to Site Editor -> Patterns
2. Create a bunch of patterns by running this in the console and reload the page:

```js
Array.from({ length: 500 }, (_, i) => i + 1).forEach((index) => {
  window.wp.apiFetch({ path: '/wp/v2/blocks', method: 'POST', data: {
    title: `Pattern ${index}`,
    content: window.wp.blocks.serialize([
      window.wp.blocks.createBlock('core/heading', { content: `Pattern ${index}` } ),
      window.wp.blocks.createBlock('core/spacer', { height: `${50 + ~~(Math.random() * 300)}px` } ),
    ]),
    status: 'publish',
  } });
});
```
3. Visit "My Patterns", scroll down and see that there's an indicator text showing how many patterns are left.
4. Try switching between different sync status
5. Try searching in the searchbox
6. Switch to different categories and try searching

### Testing Instructions for Keyboard

Follow the same steps above. The grid list is using `<Composite>` so that it should be able to navigate through arrow buttons. It's likely not accessible enough though, but that can be tackled in another PR.

## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/7753001/c3aa688b-6e12-4167-b6b1-47b50e8a882f
